src/mapping/query/function/format_number.rs

```diff
@@ -0,0 +1,254 @@
+            .collect::<Vec<String>>();
+
+        // Manipulate fractional part based on configuration.
+        match scale {
+            Some(i) if i == 0 => parts.truncate(1),
+            Some(i) => {
+                let i = i as usize;
+
+                if parts.len() == 1 {
+                    parts.push("".to_owned())
+                }
+
+                if i > parts[1].len() {
+                    for _ in 0..i - parts[1].len() {
+                        parts[1].push_str("0")
+                    }
+                } else {
+                    parts[1].truncate(i)
+                }
+            }
+            None => {}
+        }
+
+        // Manipulate integral part based on configuration.
+        if let Some(sep) = grouping_separator.as_deref() {
+            let sep = String::from_utf8_lossy(sep);
+            let start = parts[0].len() % 3;
+
+            let positions: Vec<usize> = parts[0]
+                .chars()
+                .skip(start)
+                .enumerate()
+                .map(|(i, _)| i)
+                .filter(|i| i % 3 == 0)
+                .collect();
+
+            for (i, pos) in positions.iter().enumerate() {
+                parts[0].insert_str(pos + (i * sep.len()) + start, &sep);
+            }
+        }
+
+        // Join results, using configured decimal separator.
+        Ok(Value::from(
+            parts.join(&String::from_utf8_lossy(&decimal_separator[..])),
+        ))
+    }
+
+    fn parameters() -> &'static [Parameter] {
+        &[
+            Parameter {
+                keyword: "value",
+                accepts: |v| matches!(v, Value::Integer(_) | Value::Float(_)),
+                required: true,
+            },
+            Parameter {
+                keyword: "scale",
+                accepts: |v| matches!(v, Value::Integer(_)),
+                required: false,
+            },
+            Parameter {
+                keyword: "decimal_separator",
+                accepts: |v| matches!(v, Value::Bytes(_)),
+                required: false,
+            },
+            Parameter {
+                keyword: "grouping_separator",
+                accepts: |v| matches!(v, Value::Bytes(_)),
+                required: false,
+            },
+        ]
+    }
+}
+
+impl TryFrom<ArgumentList> for FormatNumberFn {
+    type Error = String;
+
+    fn try_from(mut arguments: ArgumentList) -> Result<Self> {
+        let query = arguments.required("value")?;
+        let scale = arguments.optional("scale");
+        let decimal_separator = arguments.optional("decimal_separator");
+        let grouping_separator = arguments.optional("grouping_separator");
+
+        Ok(Self {
+            query,
+            scale,
+            decimal_separator,
+            grouping_separator,
+        })
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::mapping::query::path::Path;
+
+    #[test]
+    fn format_number() {
+        let cases = vec![
```

Comment:
:+1: I like this table-driven test approach. I did have a hard time remembering what each parameter was (especially when `None` was used) though.

Reply:
Yeah, I agree with you. I copied this from other examples, but I would love to make these tests a bit more readable.

I'll probably keep using this format for now, until all functions are implemented, and then see if there are some improvements I can make to all tests to improve the issues you've mentioned.